### PR TITLE
Feature/funcs v2

### PR DESCRIPTION
**Pull request by AymenFJA**
This pull request is adding the feature of executing pure python functions with decorators: 

``` python
from radical.pilot import PythonTask

pytask = PythonTask.pythonfunction

@pytask
def foo(a,b):
      x= a+b
      return x

cud.EXECUTABLE = foo(2,7)

```

### CODE REVIEW

File: bin/radical-pilot-agent-funcs [hunk 1]
```diff
@@ -26,7 +29,9 @@ if len(sys.argv) > 2:
 if ve and ve not in ['', 'None', None]:
 
     activate = "%s/bin/activate_this.py" % ve
-    execfile(activate, dict(__file__=activate))
+
+  # execfile(activate, dict(__file__=activate))
```
Comment (andre-merzky) [inline bin/radical-pilot-agent-funcs:33]: The old version can be removed IMHO

File: bin/radical-pilot-agent-funcs [hunk 2]
```diff
@@ -105,14 +113,15 @@ class Executor(object):
...
-            self._nw = mp.cpu_count() 
+            self._nw = mp.cpu_count()
 
         self._log.debug('#workers: %d', self._nw)
 
         self._workers = list()
         for i in range(self._nw):
             wid  = '%s.%03d' % (self._uid, i)
             proc = mp.Process(target=self._work, args=[self._uid, wid])
+            proc.daemon = True
```
Comment (andre-merzky) [inline bin/radical-pilot-agent-funcs:124]: Can you explain why that was added?  I think it explains the lingering processes we saw on the Jetstream VM.

File: bin/radical-pilot-agent-funcs [hunk 3]
```diff
@@ -193,14 +200,44 @@ class Executor(object):
...
+    #
+    def prepare_func(self, func, args):
+
+        u_ns = locals()
+        u_ns.update({'__builtins__': __builtins__})
+        function_info = {}
+        
+        function_info = pickle.loads(codecs.decode(func.encode(), "base64"))
+
+        (fn_name,fn, fn_args, fn_kwargs) = self.decompose(function_info)
```
Comment (andre-merzky) [inline bin/radical-pilot-agent-funcs:232]: The brackets are not needed, and a space is missing before `fn`.

File: bin/radical-pilot-agent-funcs [hunk 4]
```diff
@@ -193,14 +200,44 @@ class Executor(object):
...
+        u_ns.update({'__cud_args'  : fn_args,
+                     '__cud_kwargs': fn_kwargs,
+                     '__cud_output': None})
+        source = fn.split('\n')
+        fn_call = "__cud_output = {0}({1})".format(fn_name,','.join(args))
```
Comment (andre-merzky) [inline bin/radical-pilot-agent-funcs:237]: At the moment we still use the old string expansion syntax:
```py
fn_call = "__cud_output = %s(%s)" % (fn_name, ','.join(args))
```

File: bin/radical-pilot-agent-funcs [hunk 5]
```diff
@@ -217,24 +254,50 @@ class Executor(object):
...
+                    if pre.split()[0] == 'import':
+                        for mod in pre.split()[1:]:
+                            if mod not in globals():
+                                globals()[mod] = ru.import_module(mod)
+                                locals() [mod] = globals()[mod]
+
+                try:
+                    function_code = self.prepare_func(exe, args)
+                except Exception:
+                    self._log.debug("Setting the function for execution failed!.")
```
Comment (andre-merzky) [inline bin/radical-pilot-agent-funcs]: typo `!.`

File: bin/radical-pilot-agent-funcs [hunk 6]
```diff
@@ -217,24 +254,50 @@ class Executor(object):
...
+                    sys.stdout = stdout
+                    yield stdout
+                    sys.stdout = old
+                
+                with stdoutIO() as s:
+                     exec(function_code)
+        
+                #task['stdout'] = eval(cmd) # We need to find a way to keey this option!!!
+                task['stdout'] = s.getvalue()
                 task['stderr'] = None
```
Comment (andre-merzky) [inline bin/radical-pilot-agent-funcs:306]: Why don't we capture stderr also?

File: bin/radical-pilot-agent-funcs [hunk 7]
```diff
@@ -217,24 +254,50 @@ class Executor(object):
...
+                    if stdout is None:
+                       stdout = StringIO()
+                    sys.stdout = stdout
+                    yield stdout
+                    sys.stdout = old
+                
+                with stdoutIO() as s:
+                     exec(function_code)
+        
+                #task['stdout'] = eval(cmd) # We need to find a way to keey this option!!!
```
Comment (andre-merzky) [inline bin/radical-pilot-agent-funcs]: This line can go I believe.  Or if the comment is important, fix the `keey` typo (?) and remove at least two exclamation marks ;-)

File: bin/radical-pilot-agent-funcs [hunk 8]
```diff
@@ -217,24 +254,50 @@ class Executor(object):
...
+                def stdoutIO(stdout=None):
+                    old = sys.stdout
+                    if stdout is None:
+                       stdout = StringIO()
+                    sys.stdout = stdout
+                    yield stdout
+                    sys.stdout = old
+                
+                with stdoutIO() as s:
+                     exec(function_code)
```
Comment (andre-merzky) [inline bin/radical-pilot-agent-funcs]: Wouldn't this be simpler:
```
old_stdout = sys.stdout
new_stdout = None
try:
    sys.stdout = new_stdout = StringIO()
    exec(function_code)
finally:
    sys.stdout = old_stdout
task['stdout'] = new_stdout.getvalue()
```

File: bin/radical-pilot-agent-funcs (hunk 6 above)
Comment (andre-merzky) [reply bin/radical-pilot-agent-funcs:306]: On the exception check below: we could still obtain any `stdout` we captured so far.  Same for `stderr`, and then appending the exception stack maybe?  Note that RU has some [utiliy method](https://github.com/radical-cybertools/radical.utils/blob/devel/src/radical/utils/debug.py#L177) to obtain the exception stack as string.

File: examples/12_unit_function.py [hunk 9]
```diff
@@ -57,9 +57,9 @@
                    'runtime'       : 60,  # pilot runtime (min)
                    'exit_on_error' : True,
                    'project'       : config[resource].get('project', None),
-                   'queue'         : config[resource].get('queue', None),
-                   'access_schema' : config[resource].get('schema', None),
-                   'cores'         : config[resource].get('cores', 1),
+                   'queue'         : config[resource].get('queue',   None),
+                   'access_schema' : config[resource].get('schema',  None),
+                   'cores'         : 8,
```
Comment (andre-merzky) [inline examples/12_unit_function.py:62]: Please get the cores from the config file here - not all machines have 8 cores.

File: examples/12_unit_function.py [hunk 10]
```diff
@@ -76,25 +76,26 @@
         # Create a workload of ComputeUnits.
         # Each compute unit runs '/bin/date'.
 
-        n = 1024 * 2
-        report.info('create %d unit description(s)\n\t' % n)
+        n = 10
```
Comment (andre-merzky) [inline examples/12_unit_function.py:79]: 10 is boring :-P

File: examples/12_unit_function_v2.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+        # Define an [n]-core local pilot that runs for [x] minutes
+        # Here we use a dict to initialize the description object
+        pd_init = {'resource'      : resource,
+                   'runtime'       : 60,  # pilot runtime (min)
+                   'exit_on_error' : True,
+                   'project'       : config[resource].get('project', None),
+                   'queue'         : config[resource].get('queue',   None),
+                   'access_schema' : config[resource].get('schema',  None),
+                   'cores'         : 8,
```
Comment (andre-merzky) [inline examples/12_unit_function_v2.py:74]: same

File: examples/12_unit_function_v2.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+        cuds = list()
+        for i in range(0, n):
+
+            # create a new CU description, and fill it.
+            # Here we don't use dict initialization.
+            cud = rp.ComputeUnitDescription()
+            cud.pre_exec         = []
+            cud.executable       = mathma # Function reference
+            cud.arguments        = [5,7]  # Function arguments
```
Comment (andre-merzky) [inline examples/12_unit_function_v2.py:102]: space after comma

File: src/radical/pilot/compute_unit_description.py [hunk 13]
```diff
@@ -337,7 +339,7 @@ class ComputeUnitDescription(ru.Description):
     """
 
     _schema = {
-               EXECUTABLE      : str         ,
+               EXECUTABLE      : None        ,
```
Comment (andre-merzky) [inline src/radical/pilot/compute_unit_description.py:342]: I would like to discuss this change on the next devel call.  Motivation is understood, but it removes type checking from the only required task attribute...

File: src/radical/pilot/compute_unit_description.py [hunk 14]
```diff
@@ -418,6 +420,20 @@ def __init__(self, from_dict=None):
...
+           block = inspect.getsource(exe)
+
+           cu_exe_dict   = {'_cud_name':exe.__name__,
+                            '_cud_code':block,
+                            '_cud_args':'',
+                            '_cud_kwargs':''}
+
+
+           func_obj = codecs.encode(pickle.dumps(cu_exe_dict), "base64").decode()
+           self['executable'] = func_obj
```
Comment (andre-merzky) [inline src/radical/pilot/compute_unit_description.py:436]: I think we should consider `verify` optional, and to not be used in non-debug runs.  The purpose is to *verify* data, not to *transform* data.    I understand that we don't have a good place to put that code.  Maybe something like this:
```py
descr = rp.ComputeUnitDescription()
descr.function = rp.prep_function(mathma)
desc, args = [7, 8],
descr.kwargs = {'verbose'=True}
```
?  Lets discuss.  @mtitov @mturilli @iparask : any opinion?

**Review by andre-merzky — CHANGES_REQUESTED** [review]
Thanks for moving that along, Aymen!  I added a number of minor comments which should be easy to address (or to refuse! :-) ) - but also would like to discuss the API part on the next devel call (see comments at the end).

File: bin/radical-pilot-agent-funcs (hunk 2 above)
Comment (AymenFJA) [reply bin/radical-pilot-agent-funcs]: This was added by you under the branch `fix_funcs`.

Comment (andre-merzky) [reply bin/radical-pilot-agent-funcs]: Ah :-D  Let me look through that code again...

**Comment by AymenFJA** [comment]
@andre-merzky : Based on your new termination commits i started seeing the error from #2247

```
worker thread initialization failed
Traceback (most recent call last):
  File "/home/aymen/radical.pilot.sandbox/rct.func.session.js-17-185.jetstream-cloud.org.aymen.018607.0039/pilot.0000/rp_install/lib/python3.7/site-packages/radical/pilot/utils/component.py", line 472, in _worker_thread
    self._initialize()
  File "/home/aymen/radical.pilot.sandbox/rct.func.session.js-17-185.jetstream-cloud.org.aymen.018607.0039/pilot.0000/rp_install/lib/python3.7/site-packages/radical/pilot/utils/component.py", line 609, in _initialize
    self.initialize()
  File "/home/aymen/radical.pilot.sandbox/rct.func.session.js-17-185.jetstream-cloud.org.aymen.018607.0039/pilot.0000/rp_install/lib/python3.7/site-packages/radical/pilot/agent/executing/funcs.py", line 53, in initialize
    addr_wrk = self._cfg['bridges']['funcs_req_queue']
  File "/home/aymen/radical.pilot.sandbox/rct.func.session.js-17-185.jetstream-cloud.org.aymen.018607.0039/pilot.0000/rp_install/lib/python3.7/site-packages/radical/utils/munch.py", line 113, in __getitem__
    return self._data[k]
KeyError: 'bridges' 
```

File: bin/radical-pilot-agent-funcs (hunk 7 above)
Comment (AymenFJA) [reply bin/radical-pilot-agent-funcs]: This is fixed now: eval and exec can work together based on the `exe`  `obj_type`

File: bin/radical-pilot-agent-funcs (hunk 2 above)
Comment (AymenFJA) [reply bin/radical-pilot-agent-funcs]: I think this is fixed by now?

File: src/radical/pilot/compute_unit_description.py (hunk 13 above)
Comment (AymenFJA) [reply src/radical/pilot/compute_unit_description.py:342]: I missed the last development call, any update regarding this please?

**Comment by AymenFJA** [comment]
Ok, pulling and installing again with the new commits, this is working. Now we can switch between `eval` and `exec` based on the executable type provided by the user if it is an `str` or a `function reference`.

File: bin/radical-pilot-agent-funcs (hunk 2 above)
Comment (andre-merzky) [reply bin/radical-pilot-agent-funcs]: yes

File: src/radical/pilot/compute_unit_description.py (hunk 13 above)
Comment (andre-merzky) [reply src/radical/pilot/compute_unit_description.py:342]: No, that was not yet discussed.  I don't mind merging this now and to start the discussion concurrently.

**Comment by andre-merzky** [comment]
> Ok, pulling and installing again with the new commits, this is working. Now we can switch between `eval` and `exec` based on the executable type provided by the user if it is an `str` or a `function reference`.

The termination still needs take care of though before we can merge.

File: src/radical/pilot/compute_unit_description.py (hunk 14 above)
Comment (andre-merzky) [reply src/radical/pilot/compute_unit_description.py:436]: One more argument for keeping verify optional: it is costly.  It is good to have, but if we **have** to use it for millions of tasks, we give away the performance advantage we gained with our `ru.Description` implementation...

**Comment by AymenFJA** [comment]
> What happens if both `if` are not matched?

Fair, will fix it. Thanks.

**Pull request closed without merging**